esp/esp/program/models/__init__.py

```diff
@@ -787,13 +787,58 @@ def dates(self):
                 result.append(ts_day)
         return result
 
+    def datetime_range(self):
+        slots = self.getTimeSlots()
+        if slots:
+            return (min(slots).start, max(slots).end)
+        return None
+
+    # @staticmethod --- applied below after the depend_on_model call
+    @cache_function_for(60*60*24)
+    def current_program():
+        """ Guess the "current program", which is the first of the following programs that exists:
+
+        - the shortest program such that the current time is between the
+          start of its first timeslot and the end of its last timeslot
+        - the program in the future (<100 years) that will start the soonest
+        - the program in the past that ended the most recently
+        """
+        now = datetime.now()
+        far_future = now + timedelta(days=36500)
+        def currentness_penalty(program):
+            # The lower the return value (lexicographically), the more
+            # current a program is.
+            if "test" in program.name.lower():
+                return (9001, None)
+
+            datetime_range = program.datetime_range()
+            if datetime_range is None:
+                return (1337, None)
+            start, end = datetime_range
+            if start <= now <= end:
+                # most current: a program running now.
+                # tiebreak by shortest
+                return (0, (end - start))
+            elif now <= start <= far_future:
+                # second most current: program coming up in <100 years
+                # tiebreak by soonest
+                return (1, start)
+            else:
```

Comment:
This wants to be `elif start <= now` to exclude far future programs, or have another block up top similarly to `datetime_range is None`, right?